pallets/identity/src/lib.rs

```diff
@@ -95,8 +122,8 @@ decl_storage! {
         /// It stores the current identity for current transaction.
         pub CurrentDid: Option<IdentityId>;
 
-        /// (DID, claim_data, claim_issuer) -> Associated claims
-        pub Claims: double_map hasher(blake2_256) IdentityId, blake2_256(ClaimIdentifier) => IdentityClaim;
+        /// (Target ID, claim type) (issuer,scope) -> Associated claims
+        pub Claims: double_map hasher(blake2_256) Claim1stKey, blake2_256(Claim2ndKey) => IdentityClaim;
```

Comment:
It will be best if the first key only contains the target id. This is because the dApp needs to show all claims that an identity has. If the first key is just the target identity, we can easily enumerate all its claims.

Alternatively, we need to add an RPC function that loops over all claim types to fetch all claims of an identity and then return the combined result.

Reply:
I vote for option 2: Iterate over claims, because this is the performance key to optimize the following operations:

-  Fetching CDD, which is going to be a very frequency operation. In this case, first valid CDD will be enough.
-  Evaluating Claims in GTM (or any other TM) is based on rules which are defined per claim type.

Reply:
@fmiguelgarcia If I remember correctly, We don't iterate over all CDD or any other claim type in any situation on-chain. We always fetch claims from trusted providers by generating both keys and that takes O(n) time. If we move `claim_type` to `Claim2ndKey`, we'll still be able to fetch claims in O(n) time.

Reply:
@maxsam4 Actually, it is just the opposite. Let see common use-cases:
-  Fetch CDD Claims. It this case we search by `Target Identity` & `ClaimType::CDD`, and we iterate over result. e.g. in `identity/src/lib.rs` functions like `has_valid_cdd` or `fetch_cdd`.
   - `has_valid_cdd` stops at first claim which was issued by a valid CDD. There is a high probability of being the first element, since only CDD providers are allowed to create CDD claims (another task).
   - `fetch_cdd` It is similar to `has_valid_cdd` but it returns the CDD provider Id. It also has a `leeway` parameter, that we apply to Claim expiration. That's the main reason why this code is not similar to `has_valid_cdd` but their behaviours are mostly the same.
- Rules evaluation in GTM.  Each rule uses a specific `ClaimType`, so we are in the same situation, we have to fetch claims for a `Target Identity` + `ClaimType`. Optionally we can filter more using the claim `scope` and/or the `issuer` of the claim.  That optimization is in `general_tm.rs`, function `fetch_context`.

In summary, both cases use the same fist key (`Claim1stKey`), and the rule evaluation process can use even more filtering (some cases will use `Claim2ndKey`), reducing the fetching size from disk and the memory footprint.

Reply:
In none of the cases, we iterate over all results of only Claim1stKey. We always fetch claims from trusted issuer and scope. The scope is not needed in most cases but trusted issuer is needed and it will become a dos vector if we were to fetch all claims of `Claim1stKey` without trusted issuer because anyone will be able to issue a hundred claims on someone and then we'll have to iterate over all those claims to find our trusted claim. This is why we always iterate over trusted issuers and fetch single claims issued by them. The same concept applies to GTM.

I can't find any location where we are iterating solely over `Claim1stKey`. Can you please provide a specific line of code where we are doing this?

Reply:
Seems like the logic was changed in this PR. I must've missed it during the review. Due to the attack vector I mentioned above, we can't do this. We can not iterate over just Claim1stKey because it can become an unbounded array.

Reply:
@maxsam4 Unbounded claim fetch has been removed. We have **only 2 public entry points**:
-  `fetch_claim( target: IdentityId, claim_type: ClaimType, issuer: IdentityId, scope: Option<Scope> )`. It returns a `Option<IdentityClaim>`.
- `fetch_cdd( target: IdentityId, leeway: u64) -> Option<IdentityId>`. It fetches the first valid CDD claim.

**Scope** is an alias of `IdentityId`. It is an option to support claim indexation without scope (like CDD), but it is **not** a wildcard.